[clang-format] Correctly annotate C compound literal braces

Lane0218 opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/173583

clang-format mis-formats C compound literals used in macro bodies, e.g.
`#define getAddr(v, type) &(type){v}`, treating `{...}` as a block/function body
and reflowing the macro with backslashes and spaces.

This change:
- Recognizes `&( ... ){ ... }` patterns in macro bodies and marks the brace as a braced-init.
- Improves parsing stability for braced lists in macro bodies.

Tests:
- FormatTests

Lane0218 committed:
[clang-format] Keep compound literals stable in macro bodies

Fixes https://github.com/llvm/llvm-project/issues/173583.

Test: FormatTests

HazardyKnusperkeks commented on clang/lib/Format/UnwrappedLineParser.cpp:
```diff
@@ -491,6 +491,31 @@ void UnwrappedLineParser::calculateBraceTypes(bool ExpectClassBody) {
 
+  constexpr int MaxLookBack = 64;
+  const auto IsAddressOfParenExpression = [](const FormatToken *RightParen) {
+    if (!RightParen || RightParen->isNot(tok::r_paren))
```
You already checked `PrevToken` before calling this. No need for a double check.

Lane0218 committed:
[clang-format] Minimize style overrides in compound literal macro test

Test: FormatTests

Lane0218 committed:
[clang-format] Simplify preconditions for macro compound literals

NFC, addresses review feedback.

Test: FormatTests

Lane0218 commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -5858,6 +5858,26 @@ TEST_F(FormatTest, RespectWhitespaceInMacroDefinitions) {
+  Style.UseTab = FormatStyle::UT_Never;
+  Style.AlignEscapedNewlines = FormatStyle::ENAS_LeftWithLastLine;
+  Style.AllowShortFunctionsOnASingleLine = FormatStyle::SFS_Empty;
+  Style.BreakBeforeBraces = FormatStyle::BS_Attach;
```
Good point — the style overrides were not necessary to reproduce the issue. I removed the extra style settings and kept the test using only getLLVMStyle(). The test still fails without the fix and passes with the fix.

HazardyKnusperkeks commented on clang/lib/Format/UnwrappedLineParser.cpp:
```diff
@@ -528,7 +553,16 @@ void UnwrappedLineParser::calculateBraceTypes(bool ExpectClassBody) {
         }
       } else {
-        Tok->setBlockKind(BK_Unknown);
+        // In macro bodies we try to keep compound literal expressions like
```
Or maybe even add a type for the `&` and the following `(` and `{`? To make it absolutely clear?

Lane0218 replied:
Thanks for the suggestions!

I followed @owenca’s requested direction and switched to a smaller parser-side fix (in `parseParens`) rather than the loop/heuristic in `calculateBraceTypes`. I also added a `TokenAnnotatorTest` regression that checks the brace kind for `&(type){v}` is `BK_BracedInit`.

Thanks again — I agree a more general TokenAnnotator/MatchingParen-based approach could be a good follow-up refactor, but I’m trying to keep this PR minimal.

Tests: FormatTests

owenca reviewed (changes requested):
Looks like all we need is this:
```diff
diff --git a/clang/lib/Format/UnwrappedLineParser.cpp b/clang/lib/Format/UnwrappedLineParser.cpp
index c1a9161b1072..775fb9b7b3ed 100644
--- a/clang/lib/Format/UnwrappedLineParser.cpp
+++ b/clang/lib/Format/UnwrappedLineParser.cpp
@@ -2664,6 +2664,9 @@ bool UnwrappedLineParser::parseParens(TokenType AmpAmpTokenType,
           RParen->setFinalizedType(TT_TypeDeclarationParen);
         } else if (Prev->is(tok::greater) && RParen->Previous == LParen) {
           Prev->setFinalizedType(TT_TemplateCloser);
+        } else if (FormatTok->is(tok::l_brace) && Prev->is(tok::amp) &&
+                   !Prev->Previous) {
+          FormatTok->setBlockKind(BK_BracedInit);
         } else if (OptionalParens()) {
           LParen->Optional = true;
           RParen->Optional = true;
diff --git a/clang/unittests/Format/TokenAnnotatorTest.cpp b/clang/unittests/Format/TokenAnnotatorTest.cpp
index 69c2c2f17b37..a04c6bea4d05 100644
--- a/clang/unittests/Format/TokenAnnotatorTest.cpp
+++ b/clang/unittests/Format/TokenAnnotatorTest.cpp
@@ -3943,6 +3943,10 @@ TEST_F(TokenAnnotatorTest, BraceKind) {
   // Not TT_FunctionDeclarationName.
   EXPECT_TOKEN(Tokens[6], tok::kw_operator, TT_Unknown);
   EXPECT_BRACE_KIND(Tokens[9], BK_BracedInit);
+
+  Tokens = annotate("&(type){v}");
+  ASSERT_EQ(Tokens.size(), 8u) << Tokens;
+  EXPECT_BRACE_KIND(Tokens[4], BK_BracedInit);
 }
 
 TEST_F(TokenAnnotatorTest, UnderstandsElaboratedTypeSpecifier) {
```

Lane0218 committed:
[clang-format] Use verifyFormat for compound literal macro test

Drop the invalid getAddr2 macro case.

Test: FormatTests

Lane0218 committed:
[clang-format] Classify compound literal braces in macros

Handle the parser path for &(type){...} by setting the following { to BK_BracedInit.

Test: FormatTests

mydeveloperday reviewed (approved):
This looks pretty good to me, but wait for @owenca approval, but thanks for fixing this.

Lane0218 committed:
[clang-format] Remove redundant compound literal FormatTest

The compound literal macro regression is covered at the TokenAnnotator level.

Drop the extra FormatTest case and keep the coverage in TokenAnnotatorTest.

owenca committed:
Merge branch 'main' into clang-format-compound-literal
